src/NuGet.Core/NuGet.Packaging/PackageArchiveReader.cs

```diff
@@ -296,7 +296,8 @@ public override Task<bool> IsSignedAsync(CancellationToken token)
             }
 
 #if IS_DESKTOP
-            using (var reader = new BinaryReader(ZipReadStream, SigningSpecifications.Encoding, leaveOpen: true))
+            using (var bufferedStream = new ReadOnlyBufferedStream(ZipReadStream, leaveOpen: true))
```

Comment:
Why in this version we're not using `ReadOnlyBufferedStream` for `GetPrimarySignatureAsync`? Isn't it the same risk involved because we are already using it for this operation?

Reply:
https://github.com/NuGet/NuGet.Client/pull/2044 was a significant change that didn't make it in for 4.6.x.  `GetSignatureAsync(...)`/`GetPrimarySignatureAsync(...)` was [updated](https://github.com/NuGet/NuGet.Client/pull/2044/files#diff-f78aa5e2a43496c45e31b8660023d78dL243) to not use `ZipArchive`.

However, since `GetSignatureAsync(...)` still uses `ZipArchive` to check if the package is signed, I opted to leave it as is.  I think you're right that we could use `ReadOnlyBufferedStream` for `GetSignatureAsync(...)`, but we also haven't seen a perf problem with `ZipArchive`, so I'm inclined to minimize churn and to leave it as is for 4.6.x.  The perf issue was in _our_ ZIP reader.